### PR TITLE
Exclude filters

### PR DESCRIPTION
**Pull request by KhaledBousrih**

### CODE REVIEW

**Review by Souleymane-T — CHANGES_REQUESTED** [review]

File: docs/filters.md [hunk 1]
```diff
@@ -1,16 +1,19 @@
 ## Filters and ordering
 
 ### Filters
-API requests support different types of filters on the fields of the `filter_fields` declared in the datamodel:
+API requests support different types of filters on the fields of the `filter_fields` declared in the datamodel, and other specific query parameters:
 
-- **Filter Supporting Or operation:** (Applied on all fields) By adding the `__in` suffix:
+#### Filter against model fields
+
+- **Filter exact value:** (Applied on all fields within the `filter_fields`) by using a scrict equality (`field_name=field_value`) and returns all instances that have `field_value` as a value for `field_name`
```
[the remark this answers is not in the document]
Comment (KhaledBousrih) [reply docs/filters.md]: resolved, thanks

**Review by Souleymane-T — APPROVED** [review]

File: concrete_datastore/api/v1/filters.py [hunk 2]
```diff
@@ -106,6 +104,69 @@ def return_if_not_details(self, view, value, extra_condition=True):
...
+    query_parms (only ending with '!')
+    The filter method will exclude the result of super().filter_queryset
+    The exclusion must be performed on each queryparam
+    """
+
+    def get_negated_query_params(self, request, view):
+        filterset_fields = getattr(view, 'filterset_fields', ())
+        query_params = request.query_params
+        return [
+            {param[:-1]: value}
```
[the remark this answers is not in the document]
Comment (lcognat) [reply concrete_datastore/api/v1/filters.py:140]: ok got it, str without last char

File: concrete_datastore/api/v1/filters.py [hunk 3]
```diff
@@ -230,63 +322,71 @@ def get_schema_operation_parameters(self, view):
...
         params = [
             {
-                'name': f'{field_name}__in',
+                'name': f'{field_name}__in{neg}',
                 'required': False,
                 'in': 'query',
-                'description': 'List of values separated by comma',
+                'description': (
+                    'List of values separated by comma'
+                    f'{" (to exclude)" if neg=="!" else ""}'
```
Comment (lcognat) [inline concrete_datastore/api/v1/filters.py:330]: spacing should be corrected

File: concrete_datastore/api/v1/filters.py [hunk 4]
```diff
@@ -295,26 +395,35 @@ class FilterSupportingContainsBackend(
     def get_schema_operation_parameters(self, view):
         params = [
             {
-                'name': f'{field_name}__contains',
+                'name': f'{field_name}__contains{neg}',
                 'required': False,
+                'description': f'String{" (to exclude)" if neg=="!" else ""}',
```
Comment (lcognat) [inline concrete_datastore/api/v1/filters.py:400]: spacing should be corrected

File: concrete_datastore/api/v1/filters.py [hunk 5]
```diff
@@ -434,16 +563,20 @@ class FilterSupportingRangeBackend(
...
         params = [
             {
-                'name': f'{field_name}__range',
+                'name': f'{field_name}__range{neg}',
                 'required': False,
                 'in': 'query',
-                'description': 'A range of values separated by comma',
+                'description': (
+                    'A range of values separated by comma'
+                    f'{" (to exclude)" if neg=="!" else ""}'
```
Comment (lcognat) [inline concrete_datastore/api/v1/filters.py:571]: spacing should be corrected

File: concrete_datastore/api/v1/filters.py [hunk 6]
```diff
@@ -544,16 +685,20 @@ def get_schema_operation_parameters(self, view):
...
                 [
                     {
-                        'name': f'{field_name}__{key}',
+                        'name': f'{field_name}__{key}{neg}',
                         'required': False,
                         'in': 'query',
-                        'description': description,
+                        'description': (
+                            f'{description}'
+                            f'{" (to exclude)" if neg=="!" else ""}'
```
Comment (lcognat) [inline concrete_datastore/api/v1/filters.py:693]: spacing should be corrected

File: concrete_datastore/api/v1/filters.py [hunk 7]
```diff
@@ -612,26 +771,36 @@ class FilterSupportingForeignKey(
     def get_schema_operation_parameters(self, view):
         params = [
             {
-                'name': f'{field_name}_uid',
+                'name': f'{field_name}_uid{neg}',
                 'required': False,
                 'in': 'query',
-                'description': 'UID of the FK',
+                'description': (
+                    f'UID of the FK{" (to exclude)" if neg=="!" else ""}'
```
Comment (lcognat) [inline concrete_datastore/api/v1/filters.py:778]: spacing should be corrected

**Review by lcognat — CHANGES_REQUESTED** [review]

File: concrete_datastore/api/v1/filters.py [hunk 8]
```diff
@@ -125,15 +186,19 @@ def remove_from_queryset(self, view):
...
         params = [
             {
-                'name': f'{field_name}__distance',
+                'name': f'{field_name}__distance{neg}',
                 'required': False,
                 'in': 'query',
-                'description': 'DISTANCE,LONGITUDE,LATITUDE',
+                'description': (
+                    'DISTANCE,LONGITUDE,LATITUDE'
+                    f'{" (to exclude)"if neg=="!" else ""}'
```
[the remark this answers is not in the document]
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py]: corrected

File: concrete_datastore/api/v1/filters.py (hunk 3 above)
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py]: corrected

File: concrete_datastore/api/v1/filters.py (hunk 5 above)
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py]: corrected

File: concrete_datastore/api/v1/filters.py (hunk 7 above)
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py]: corrected

File: concrete_datastore/api/v1/filters.py (hunk 6 above)
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py]: corrected

File: concrete_datastore/api/v1/filters.py (hunk 4 above)
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py]: corrected

File: concrete_datastore/api/v1/filters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/filters.py:140]: yes, to remove the `!` at the end of the param

**Pull request merged**